nupic/torch/modules/sparse_weights.py

```diff
@@ -136,33 +123,27 @@ class SparseWeights(SparseWeightsBase):
     """
 
     def __init__(self, module, weight_sparsity=None, sparsity=None):
-        assert isinstance(module, nn.Linear)
-        assert 0 < (weight_sparsity or sparsity) < 1
+        assert len(module.weight.shape) == 2, "Should resemble a nn.Linear"
+        assert 0 <= (weight_sparsity or sparsity) <= 1
```

Comment:
This assert was placed to prevent the creation of `SparseWeight` layers with 0% or 100% sparsity. Both case should be considered to be a bug.
- `SparseWeight` with `sparsity=0` is not sparse, it similar to `torch.nn.Identity` but using more memory to hold the dense mask.
- `SparseWeight` with `sparsity=1` is not sparse, it is the same as replacing the layer with zeros.

These edge cases make sense in the `PrunableSparseWeights` context where the sparsity may be changed during training but not in the static network context where the sparsity is fixed at initialization.

See discussion in https://github.com/numenta/nupic.torch/pull/67

Reply:
Can I add a `allow_extremes=False` bool to the `__init__` that allows these values? The problem with having these asserts is that every subclass needs to be careful not to call `SparseWeightsBase.__init__`, which makes `super()` code really complicated. If subclasses have the option of setting `allow_extremes=True`, it would solve that problem.

Reply:
Yes, you could add `allow_extremes=False` to `SparseWeightsBase` but the leaving the assert on `SparseWeight` and  `SparseWeight2d` is a good compromise